File: app/adapters/openai_compat.py
"""OpenAI-compatible chat completions adapter.

Handles two flavors of the same wire format:
  - OpenWebUI's proxy:  POST {base}/api/chat/completions
  - Standard OpenAI-compatible servers (vLLM, LM Studio, TGI, Ollama's own
    compat shim): POST {base}/v1/chat/completions

Both speak SSE ("data: {...}\\n\\n" lines, terminated by "data: [DONE]") with
delta.content chunks and an optional trailing usage object. We try the
OpenWebUI path first (since that's the documented primary use case here)
and fall back to /v1/ on a 404, so either server type works without extra
configuration.
"""
from __future__ import annotations

import json
from typing import AsyncIterator

import httpx

from app.adapters.base import Adapter
from app.models import Endpoint, PromptCase, StreamEvent

CANDIDATE_PATHS = ["/api/chat/completions", "/v1/chat/completions"]


def _headers(endpoint: Endpoint) -> dict[str, str]:
    headers = {"Content-Type": "application/json"}
    if endpoint.api_key:
        headers["Authorization"] = f"Bearer {endpoint.api_key}"
    return headers


def _payload(endpoint: Endpoint, case: PromptCase) -> dict:
    messages = []
    if case.system_prompt:
        messages.append({"role": "system", "content": case.system_prompt})
    messages.append({"role": "user", "content": case.user_prompt})
    payload: dict = {
        "model": endpoint.model,
        "messages": messages,
        "stream": True,
        "temperature": case.temperature,
    }
    if case.max_tokens is not None:
        payload["max_tokens"] = case.max_tokens
    if case.seed is not None:
        payload["seed"] = case.seed
    return payload


class OpenAICompatAdapter(Adapter):
    async def stream_chat(
        self,
        client: httpx.AsyncClient,
        endpoint: Endpoint,
        case: PromptCase,
    ) -> AsyncIterator[StreamEvent]:
        base = endpoint.base_url.rstrip("/")
        headers = _headers(endpoint)
        payload = _payload(endpoint, case)

        last_error: Exception | None = None
        for i, path in enumerate(CANDIDATE_PATHS):
            url = base + path
            try:
                async with client.stream("POST", url, headers=headers, json=payload) as resp:
                    if resp.status_code == 404 and i < len(CANDIDATE_PATHS) - 1:
                        last_error = httpx.HTTPStatusError(
                            "404, trying next path", request=resp.request, response=resp
                        )
                        continue
                    if resp.status_code >= 400:
                        body = await resp.aread()
                        yield StreamEvent(
                            kind="error",
                            error_message=f"HTTP {resp.status_code}: {body[:500]!r}",
                        )
                        return

                    prompt_tokens: int | None = None
                    completion_tokens: int | None = None
                    async for line in resp.aiter_lines():
                        if not line or not line.startswith("data:"):
                            continue
                        data = line[len("data:"):].strip()
                        if data == "[DONE]":
                            break
                        try:
                            obj = json.loads(data)
                        except json.JSONDecodeError:
                            continue

                        usage = obj.get("usage")
                        if usage:
                            prompt_tokens = usage.get("prompt_tokens", prompt_tokens)
                            completion_tokens = usage.get("completion_tokens", completion_tokens)

                        choices = obj.get("choices") or []
                        if choices:
                            delta = choices[0].get("delta") or {}
                            text = delta.get("content") or ""
                            if text:
                                yield StreamEvent(kind="delta", text=text)

                    yield StreamEvent(
                        kind="done",
                        prompt_tokens=prompt_tokens,
                        completion_tokens=completion_tokens,
                    )
                    return
            except httpx.HTTPError as exc:
                last_error = exc
                continue

        yield StreamEvent(kind="error", error_message=f"request failed: {last_error}")

    async def list_models(self, client: httpx.AsyncClient, endpoint: Endpoint) -> list[str]:
        base = endpoint.base_url.rstrip("/")
        headers = _headers(endpoint)
        last_error: Exception | None = None
        for path in ["/api/models", "/v1/models"]:
            try:
                resp = await client.get(base + path, headers=headers)
                if resp.status_code == 404:
                    continue
                resp.raise_for_status()
                data = resp.json()
                items = data.get("data") or data.get("models") or []
                names = []
                for item in items:
                    if isinstance(item, str):
                        names.append(item)
                    elif isinstance(item, dict):
                        names.append(item.get("id") or item.get("name") or "")
                return [n for n in names if n]
            except httpx.HTTPError as exc:
                last_error = exc
                continue
        raise RuntimeError(f"could not list models: {last_error}")

File: lib/adapters/openaiCompat.ts
// OpenAI-compatible chat completions adapter.
//
// Handles two flavors of the same wire format:
//   - OpenWebUI's proxy:  POST {base}/api/chat/completions
//   - Standard OpenAI-compatible servers (vLLM, LM Studio, TGI, Ollama's
//     own compat shim): POST {base}/v1/chat/completions
//
// Both speak SSE ("data: {...}\n\n" lines, terminated by "data: [DONE]")
// with delta.content chunks and an optional trailing usage object. We try
// the OpenWebUI path first and fall back to /v1/ on a 404, so either
// server type works without extra configuration.

import type { Adapter } from "@/lib/adapters/base";
import { authHeaders, streamLines } from "@/lib/adapters/base";
import type { Endpoint, PromptCase, StreamEvent } from "@/lib/types";

const CANDIDATE_CHAT_PATHS = ["/api/chat/completions", "/v1/chat/completions"];
const CANDIDATE_MODEL_PATHS = ["/api/models", "/v1/models"];

function buildPayload(endpoint: Endpoint, promptCase: PromptCase) {
  const messages: { role: string; content: string }[] = [];
  if (promptCase.system_prompt) messages.push({ role: "system", content: promptCase.system_prompt });
  messages.push({ role: "user", content: promptCase.user_prompt });

  const payload: Record<string, unknown> = {
    model: endpoint.model,
    messages,
    stream: true,
    temperature: promptCase.temperature,
  };
  if (promptCase.max_tokens != null) payload.max_tokens = promptCase.max_tokens;
  if (promptCase.seed != null) payload.seed = promptCase.seed;

  // OpenAI-compatible structured output. Whether the backend behind
  // OpenWebUI's proxy actually honors this (vs. silently dropping it) is
  // exactly the kind of thing worth testing here rather than assuming --
  // a schema that fails to parse is dropped rather than sent malformed.
  if (promptCase.json_schema) {
    try {
      const schema = JSON.parse(promptCase.json_schema);
      payload.response_format = { type: "json_schema", json_schema: { name: "response", schema, strict: true } };
    } catch {
      // ignore -- request proceeds without a format constraint
    }
  }

  return payload;
}

async function* streamChat(endpoint: Endpoint, promptCase: PromptCase): AsyncGenerator<StreamEvent> {
  const base = endpoint.base_url.replace(/\/+$/, "");
  const headers = authHeaders(endpoint);
  const payload = buildPayload(endpoint, promptCase);

  let lastError = "";
  for (let i = 0; i < CANDIDATE_CHAT_PATHS.length; i++) {
    const url = base + CANDIDATE_CHAT_PATHS[i];
    let resp: Response;
    try {
      resp = await fetch(url, { method: "POST", headers, body: JSON.stringify(payload) });
    } catch (exc) {
      lastError = `request failed: ${exc}`;
      continue;
    }

    if (resp.status === 404 && i < CANDIDATE_CHAT_PATHS.length - 1) {
      lastError = "404, trying next path";
      continue;
    }
    if (resp.status >= 400) {
      const body = await resp.text();
      yield { kind: "error", error_message: `HTTP ${resp.status}: ${body.slice(0, 500)}` };
      return;
    }

    let promptTokens: number | null = null;
    let completionTokens: number | null = null;

    for await (const line of streamLines(resp)) {
      if (!line.startsWith("data:")) continue;
      const data = line.slice("data:".length).trim();
      if (data === "[DONE]") break;
      let obj: any;
      try {
        obj = JSON.parse(data);
      } catch {
        continue;
      }

      const usage = obj.usage;
      if (usage) {
        promptTokens = usage.prompt_tokens ?? promptTokens;
        completionTokens = usage.completion_tokens ?? completionTokens;
      }

      const choice = obj.choices?.[0];
      const text = choice?.delta?.content;
      if (text) yield { kind: "delta", text };
    }

    yield {
      kind: "done",
      prompt_tokens: promptTokens,
      completion_tokens: completionTokens,
      server_load_s: null,
      server_prompt_eval_s: null,
      server_eval_s: null,
    };
    return;
  }

  yield { kind: "error", error_message: `request failed: ${lastError}` };
}

async function listModels(endpoint: Endpoint): Promise<string[]> {
  const base = endpoint.base_url.replace(/\/+$/, "");
  const headers = authHeaders(endpoint);
  let lastError = "";
  for (const path of CANDIDATE_MODEL_PATHS) {
    try {
      const resp = await fetch(base + path, { headers });
      if (resp.status === 404) continue;
      if (!resp.ok) {
        lastError = `HTTP ${resp.status}`;
        continue;
      }
      const data = await resp.json();
      const items: any[] = data.data ?? data.models ?? [];
      const names = items
        .map((item) => (typeof item === "string" ? item : item.id ?? item.name ?? ""))
        .filter(Boolean);
      return names;
    } catch (exc) {
      lastError = String(exc);
    }
  }
  throw new Error(`could not list models: ${lastError}`);
}

export const openaiCompatAdapter: Adapter = { streamChat, listModels };
